sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/AssemblyReference.cs

```diff
@@ -18,62 +18,93 @@
     {
-        internal class NuGetInfo : IEquatable<NuGetInfo>
-        {
-            internal const string LatestVersion = null;
+        internal const string NuGetLatestVersion = null;
+
+        private const string PackagesFolderRelativePath = @"..\..\..\..\packages\";
+
+        private static readonly string systemAssembliesFolder =
+            new FileInfo(typeof(object).Assembly.Location).Directory.FullName;
+
+        private static readonly PackageManager packageManager =
+            new PackageManager(CreatePackageRepository(), PackagesFolderRelativePath);
 
-            public string Name { get; }
-            public SemanticVersion Version { get; }
+        private static readonly Dictionary<string, MetadataReference[]> metadataReferenceCache =
+            new Dictionary<string, MetadataReference[]>();
 
-            internal NuGetInfo(string id, string version = LatestVersion)
+        private static readonly List<string> allowedNugetLibDirectories =
+            new List<string>
             {
-                Name = id;
-                Version = SemanticVersion.ParseOptionalVersion(version);
-            }
+                "lib",
+                "portable-net45",
+                "net40",
+                "net45",
+                "netstandard1.0",
+                "netstandard1.1",
+                "netstandard2.0"
+            };
 
-            public override bool Equals(object obj) =>
-                Equals(obj as NuGetInfo);
+        private static IPackageRepository CreatePackageRepository()
+        {
+            var currentFolder = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            var localSettings = Settings.LoadDefaultSettings(new PhysicalFileSystem(currentFolder), null, null);
 
-            public override int GetHashCode() =>
-                $"{Name}|{Version}".GetHashCode();
+            // Get a package source provider that can use the settings
+            var packageSourceProvider = new PackageSourceProvider(localSettings);
 
-            public bool Equals(NuGetInfo other) =>
-                other != null &&
-                other.Name == Name &&
-                other.Version == Version;
-        }
+            // Create an aggregate repository that uses all of the configured sources
+            var aggregateRepository = packageSourceProvider.CreateAggregateRepository(PackageRepositoryFactory.Default,
+                true /* ignore failing repos. Errors will be logged as warnings. */ );
 
-        public string Name { get; }
-        public NuGetInfo NuGet { get; }
+            return aggregateRepository;
+        }
 
-        private AssemblyReference(string assemblyName, NuGetInfo nuGetInfo = null)
+        public static MetadataReference FromFrameworkAssembly(string assemblyName)
         {
-            Name = assemblyName;
-            NuGet = nuGetInfo;
+            return metadataReferenceCache.GetOrAdd(assemblyName, ProcessAssembly).Single();
```

Comment:
Could the cache entry contain zero or more than one items? Why not convert to non-array entries?

Reply:
I am not expecting something else than one + if that's the case I want to know about it hence this failure.

Reply:
Please, add a proper check with a descriptive error message then. It would be much easier for the developer who will encounter the Linq exception after several months.